### ISSUE TITLE
Doom64EX support (working directory issue)

### ISSUE THREAD

**Issue by DanilaSpevak** [Youda008/DoomRunner#31]
I tried to add Doom64EX as an engine to make DoomRunner the launcher of all my old Doom, and it's not working correctly. The launcher generates correct command line but the issue is with the working directory. When run from DoomRunner, the Current directory is DoomRunner's, not the one where Doom64ex is located.

![2021-01-03_13h21_36](https://user-images.githubusercontent.com/42034406/103474589-40725f00-4dc7-11eb-87cc-63d70d130a4b.png)

So it can't find it's own files. I think that might happen to some other older sourceports as well. Wouldn't it make more sense generally to run engine processes using their location as a working directory?

**Comment by Youda008** [comment]
Hm, i have never thought about it. Until now all paths in the generated command have been relative to the DoomRunner's directory and there were no issues. But if some engines have problems with it, it makes sense to do it as you say.
________________________________

**Comment by Berzasperd** [comment]
This seems to be related to a similar issue I had with using DoomRunner with Crispy Doom.  When launched using DoomRunner, Crispy Doom assumes that it is running in the same directory that the DoomRunner.exe is located in.  As a result, it will create its own config files there and read from those, ignoring the ones in crispy-doom.exe's home directory.  I haven't figured out which program is causing this issue yet, as it does not seem to affect other source ports, like GZDoom.

On the topic of Doom64EX, I believe that running DOOM64.EXE without supplying it any parameters will have the program look for DOOM64.WAD in the directory that DOOM64.EXE believes that it itself is in.  In DanilaSpevak's case, launching Doom64EX through DoomRunner causes DOOM64.EXE to think the files it needs to look for are in the same directory as DoomRunner, which for certain users may or may not be the case.

Workarounds:  DOOM64.EXE accepts parameters similar to those of other Doom source ports.  In particular, it will accept -iwad followed by the file path and name in quotation marks.  An example would be,

Additional command line arguments (preset) [ -iwad "sourceports/doom64ex/DOOM64.WAD" ] 

// brackets represent the fillable field.  do not include leading or trailing space.  file path in quotation marks should match
// the path to your wad file.  this also works if you copy the .wad file into the directory where DoomRunner lists IWADs
// and select it from that list, as DoomRunner will automatically generate the necessary -iwad flag

For myself, I have placed all of my iwads into a folder by themselves, so that I can have DoomRunner automatically list them and let me select them for each preset.  I have tested DoomRunner's ability to launch other source ports, such as those for Quake and Hexen 2.  These source ports behave similarly to Danila's issue, in which they look for their own files in DoomRunner's directory and not their own.  The workaround for these is to pass the -basedir or +set basedir parameters to the source ports.  The ability for the source port to look for files in other directories seems to be specific to the source port program.

A lot of these things come down to how the end user wants to organize their files in relation to DoomRunner.  In my experience, GZDoom will take anything you throw at it, so long as you supply a proper -iwad parameter when launching it.  Cripsy Doom gets confused and puts its config files in the wrong spot, so I created a separate copy of DoomRunner.exe, renamed it, and placed it in the directory with crispy-doom.exe.  This way I can have Crispy Doom specific presets within DoomRunner, at the cost of having to make another shortcut specifically for Crispy Doom related things.

In summary, my tests have suggested that running certain source ports by launching them from DoomRunner.exe will confuse the source port into thinking it is also in DoomRunner's directory, regardless of specifying the actual path in DoomRunner's parameters or settings.  This will cause Danila's issue in which the launched source port can't automatically find iwads in its directory because it is not looking in its directory for them.  The current workaround is to make sure that the iwad path is specified by the -iwad parameter at all times.  This may cause some users to resort to creating duplicates of the .wad or .exe files.  Hope the extra information helps.  I can draw up a diagram of my folder and file structure for reference if that would also help.

**Comment by Berzasperd** [comment]
Okay, did a little more testing and found some potentially useful details.

Source ports that accept the -config parameter can be told exactly which config file to use during launch, and will not create new ones in DoomRunner.exe's directory.  Doom64EX and Chocolate/Crispy Doom support this.  The difference is that these source ports use the .cfg file extension, which DoomRunner does not support in its drop down menu.

Workaround: Manually type in -config "filepath/config.cfg" for your preset to use the specified file consistently and not generate new files.

Consequences: If a user launches one source port that creates and reads "config.cfg" and then launches a different source port that also creates and reads the same file name, the second source port may overwrite the first config file or fail due to an error reading that file.  This occurs if the user does not type in their own -config parameter.  I have not tested this situation, since Doom64EX and Crispy Doom have different names for their config files.

Potential Fix 1: Make DoomRunner tell the source port that it is launching to stay within its own directory rather than using DoomRunner's directory to look for files.  The relative file pathing works very well for command line parameters, but lacking those, some source ports seem to be easily confused.  I am not sure how viable this fix is.  I can't tell if this is a result of DoomRunner, or how Windows interprets programs launching other programs.

Potential Fix 2: Add support for .cfg files in DoomRunner's "Select config (optional)" drop down menu.  This would allow users to more easily select the config file in question without having to type it into the additional arguments field.  This would be a nice addition, though it will not solve the issue if the user leaves the config file blank.  There may still be other unseen issues regarding a source port looking outside of its own directory, but I am not aware of any others at this time.

**Comment by Youda008** [comment]
Thank you for additional usecases. In that case DoomRunner will be reworked to switch the current dir to the engine's one before launch.

**Comment by Youda008** [comment]
Alright, this has been fixed. It will be part of next release.
